### PR TITLE
Task/um 837 update tutorials examples

### PR DESCRIPTION
**Pull request by kab163**
Trying to remove any unnecessary DynamicPoolMap instances used in documentation, examples, tests, and tools.

### Known instances of DynamicPoolMap:
(updated 6/3)
- benchmarks/pool_stress_test.cpp
- src/umpire/Replay.cpp
- src/umpire/interface/c_fortran/wrapResourceManager.cpp*
- src/umpire/interface/yaml/umpire_types.yaml*
- tests/applications/poison_memory_checks.cpp*
- tests/applications/pool_allocator_stress.cpp*
- tests/integration/operation_tests.cpp*
- tests/integration/primary_pool_performance_tests.cpp*
- tests/integration/primary_pool_tests.cpp*
- tests/integration/strategy_tests.cpp*
- tools/replay/ReplayInterpreter.cpp*
- tools/replay/ReplayOperationManager.cpp*
- src/umpire/strategy/DynamicPool.hpp
- src/umpire/strategy/DynamicPoolMap.cpp
- src/umpire/strategy/DynamicPoolMap.hpp

(* means it's been vetted and files added in cmake to handle it)

### Known instances of DynamicPool:
(updated 5/19)
- src/umpire/interface/c_fortran/wrapfUmpire_strategy.f **

### CODE REVIEW

File: src/umpire/strategy/DynamicPool.hpp [hunk 1]
```diff
@@ -7,12 +7,12 @@
...
 #define UMPIRE_DynamicPool_HPP
 
-#include "umpire/strategy/DynamicPoolMap.hpp"
+#include "umpire/strategy/DynamicPoolList.hpp"
 
 namespace umpire {
 namespace strategy {
 
-using DynamicPool = DynamicPoolMap;
+using DynamicPool = DynamicPoolList;
```
Comment (davidbeckingsale) [inline src/umpire/strategy/DynamicPool.hpp:15]: Please revert this change.

File: examples/tutorial/tut_replay_log.json [hunk 2]
```diff
@@ -15,8 +15,8 @@
 { "kind":"replay", "uid":27494, "timestamp":1558388052595814979, "event": "makeMemoryResource", "payload": { "name": "DEVICE_CONST" }, "result": "0x101d7bb0" }
 # _sphinx_tag_doc_makememoryresource_end
 # _sphinx_tag_doc_makeallocator_start
-{ "kind":"replay", "uid":27494, "timestamp":1558388052595864262, "event": "makeAllocator", "payload": { "type":"umpire::strategy::DynamicPool", "with_introspection":true, "allocator_name":"pool", "args": [ "HOST" ] } }
-{ "kind":"replay", "uid":27494, "timestamp":1558388052595903505, "event": "makeAllocator", "payload": { "type":"umpire::strategy::DynamicPool", "with_introspection":true, "allocator_name":"pool", "args": [ "HOST" ] }, "result": { "allocator_ref":"0x108a8730" } }
```
Comment (kab163) [inline examples/tutorial/tut_replay_log.json:19]: This is probably not needed lol (maybe I was just a little too excited here haha)

File: examples/tutorial/tut_replay_log.json [hunk 3]
```diff
@@ -15,8 +15,8 @@
...
+{ "kind":"replay", "uid":27494, "timestamp":1558388052595864262, "event": "makeAllocator", "payload": { "type":"umpire::strategy::QuickPool", "with_introspection":true, "allocator_name":"pool", "args": [ "HOST" ] } }
+{ "kind":"replay", "uid":27494, "timestamp":1558388052595903505, "event": "makeAllocator", "payload": { "type":"umpire::strategy::QuickPool", "with_introspection":true, "allocator_name":"pool", "args": [ "HOST" ] }, "result": { "allocator_ref":"0x108a8730" } }
```
Comment (kab163) [inline examples/tutorial/tut_replay_log.json:19]: This is probably not needed lol (maybe I was just a little too excited here haha)

Comment (mcfadden8) [reply examples/tutorial/tut_replay_log.json:19]: This look great!

**Comment by kab163** [comment]
https://umpire.readthedocs.io/en/task-um-837-update-tutorials-examples/cookbook.html

File: .gitlab/lassen-jobs.yml [hunk 4]
```diff
@@ -75,18 +75,18 @@ gcc_8_3_1_cuda:
...
   variables:
-    SPEC: "+fortran+cuda %pgi@19.10 cxxflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 cflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 ^cuda@10.1.168"
+    SPEC: "+fortran+cuda %pgi@19.10 cxxflags='--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 -w' cflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 ^cuda@10.1.168"
   allow_failure: true
   extends: .build_and_test_on_lassen
 
 pgi_20_4_gcc_8_3_1_cuda_11:
   variables:
-    SPEC: "+fortran+cuda %pgi@20.4 cxxflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 cflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 ^cuda@11.0.2"
+    SPEC: "+fortran+cuda %pgi@20.4 cxxflags='--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 -w' cflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 ^cuda@11.0.2"
```
Comment (mcfadden8) [inline .gitlab/lassen-jobs.yml:89]: What do these changes have to do with updating the tutorial examples?  They seem more general in nature.  I suggest that these be broken out to their own PR.

File: .gitlab/ruby-jobs.yml [hunk 5]
```diff
@@ -36,13 +36,13 @@ icpc_19_1_0:
...
   variables:
-    SPEC: "+fortran %pgi@19.7 cxxflags=-I/usr/tce/packages/gcc/gcc-4.9.3/include/c++/4.9.3"
+    SPEC: "+fortran %pgi@19.7 cxxflags='-I/usr/tce/packages/gcc/gcc-4.9.3/include/c++/4.9.3 -Minform=severe'"
     PGI_LOCALRC: /collab/usr/global/tools/pgi/etc/localrc-gcc-4.9.3
   extends: .build_and_test_on_ruby
 
 pgi_20_1:
   variables:
-    SPEC: "+fortran %pgi@20.1"
+    SPEC: "+fortran %pgi@20.1 cxxflags=-Minform=severe"
```
Comment (mcfadden8) [inline .gitlab/ruby-jobs.yml:45]: Same comment as in lassen-jobs.yml

File: .gitlab/lassen-jobs.yml (hunk 4 above)
Comment (davidbeckingsale) [reply .gitlab/lassen-jobs.yml:89]: This PR includes the deprecation. This is to avoid warnings from that, that will become errors during CI builds. These extra flags should be handled in `cmake/SetupCompilerFlags.cmake` like they are for the other compilers.

File: src/umpire/interface/umpire_shroud.yaml [hunk 6]
```diff
@@ -130,6 +136,10 @@ declarations:
         options:
           F_create_generic: False
         cxx_template:
+        - instantiation: <umpire::strategy::DynamicPoolMap>
+          format:
+            template_suffix: _map_pool
```
Comment (davidbeckingsale) [inline src/umpire/interface/umpire_shroud.yaml:141]: This should just be "_pool" to avoid breaking the C/Fortran API.

Comment (kab163) [reply src/umpire/interface/umpire_shroud.yaml:141]: whoopsies.. i thought i was being cool lol i'll change that

File: .gitlab/lassen-jobs.yml [hunk 7]
```diff
@@ -75,18 +75,18 @@ gcc_8_3_1_cuda:
 
 pgi_19_10_cuda:
   variables:
-    SPEC: "+fortran+cuda %pgi@19.10 ^cuda@10.1.168"
+    SPEC: "+fortran+cuda %pgi@19.10 ^cuda@11.0.2"
```
Comment (davidbeckingsale) [inline .gitlab/lassen-jobs.yml:78]: If you switch this back to cuda 10 it should pass.

Comment (kab163) [reply .gitlab/lassen-jobs.yml]: (was already in progress hehe i ran the tests and it passes locally at least! fingers crossed lol)

File: tests/integration/CMakeLists.txt [hunk 8]
```diff
@@ -220,3 +222,20 @@ if (ENABLE_C OR ENABLE_FORTRAN)
 endif()
 
 add_subdirectory(io)
+
+if (ENABLE_WARNINGS_AS_ERRORS)
```
Comment (davidbeckingsale) [inline tests/integration/CMakeLists.txt:226]: Either remove this guard, or change it to check if the compiler is MSVC also.

File: .gitlab/lassen-jobs.yml [hunk 9]
```diff
@@ -80,13 +80,13 @@ pgi_19_10_cuda:
 
 pgi_19_10_gcc_8_3_1_cuda:
   variables:
-    SPEC: "+fortran+cuda %pgi@19.10 cxxflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 cflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 ^cuda@10.1.168"
+    SPEC: "+fortran+cuda %pgi@19.10 cxxflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 cflags=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 cuda_arch=70 ^cuda@11.0.2"
```
Comment (davidbeckingsale) [inline .gitlab/lassen-jobs.yml:83]: Can this be CUDA 10 again?

Comment (mcfadden8) [reply .gitlab/lassen-jobs.yml:83]: Is this a request to change it back to CUDA 10 or just whether it can be?

File: CHANGELOG.md [hunk 10]
```diff
@@ -92,6 +92,10 @@ Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to
 
 - Replay testing disabled during HIP builds with `ENABLE_TOOLS`=On
 
+- `DynamicPoolMap` marked deprecated. `QuickPool` should be used instead.
+
+- Changed most uses of DynamicPoolMap to QuickPool.
```
Comment (mcfadden8) [inline CHANGELOG.md:97]:
```suggestion
- Changed most internal and test uses of DynamicPoolMap to QuickPool.
```

File: docs/sphinx/conf.py [hunk 11]
```diff
@@ -31,7 +31,7 @@
 read_the_docs_build = os.environ.get('READTHEDOCS', None) == 'True'
 
 if read_the_docs_build:
-    print('Running on read the docs')
+    print ('Running on read the docs')
```
Comment (mcfadden8) [inline docs/sphinx/conf.py:34]:
```suggestion
    print('Running on read the docs')
```

File: docs/sphinx/cookbook/dynamic_pool_heuristics.rst [hunk 12]
```diff
@@ -1,11 +1,11 @@
 .. _dynamic_pool_heuristics:
 
 ===========================================================
-Improving DynamicPool Performance with a Coalesce Heuristic
+Improving DynamicPoolList Performance with a Coalesce Heuristic
```
Comment (mcfadden8) [inline docs/sphinx/cookbook/dynamic_pool_heuristics.rst:4]: Is there already a `QuickPool` document for this section?  If not, I suggest that we change this to be more generic.  Somthing like:
```suggestion
Improving Pool Performance with a Coalesce Heuristic
```
We could consider renaming this entire cookbook section and provide code examples for both QuickPool and DynamicPoolList.  We could also talk about why some heuristics are better for QuickPool than for DynamicPoolList.  I would be fine with not changing this now if we could get it documented in a Jira ticket.

File: .gitlab/lassen-jobs.yml (hunk 9 above)
Comment (kab163) [reply .gitlab/lassen-jobs.yml:83]: A request lol at one point in time, I thought I needed to change these to cuda 11 so I could use the "--forward-unknown-to-host-compiler" flag so that nvcc would stop complaining about not knowing what PGI's "-Minform=severe" flag is. However, later on we went back and changed the way that I pass the PGI flags in and it become unnecessary to keep around.

Comment (kab163) [reply .gitlab/lassen-jobs.yml:83]: and I'll probably also undo the cuda_arch change too. That should be part of a separate PR that has the device allocator improvements.

File: src/umpire/interface/c_fortran/typesUmpire.h [hunk 13]
```diff
@@ -107,19 +107,19 @@ struct s_umpire_strategy_allocationprefetcher {
 };
 typedef struct s_umpire_strategy_allocationprefetcher umpire_strategy_allocationprefetcher;
 
-// helper capsule_umpire_strategy_dynamicpool
-struct s_umpire_strategy_dynamicpool {
+// helper capsule_umpire_strategy_dynamicpoollist
+struct s_umpire_strategy_dynamicpoollist {
```
Comment (mcfadden8) [inline src/umpire/interface/c_fortran/typesUmpire.h:111]: @kab163 and @davidbeckingsale, I don't think we can do this because it will cause C and Fortran applicaion builds to fail

File: src/umpire/interface/c_fortran/wrapResourceManager.cpp [hunk 14]
```diff
@@ -97,7 +97,8 @@ umpire_allocator * umpire_resourcemanager_make_allocator_pool(
     umpire::Allocator * SHCXX_allocator =
         static_cast<umpire::Allocator *>(allocator.addr);
     umpire::Allocator * SHCXX_rv = new umpire::Allocator;
-    *SHCXX_rv = SH_this->makeAllocator<umpire::strategy::DynamicPool>(
+    *SHCXX_rv =
+        SH_this->makeAllocator<umpire::strategy::DynamicPoolMap>(
```
Comment (mcfadden8) [inline src/umpire/interface/c_fortran/wrapResourceManager.cpp:101]: We need to be careful with the C/Fortran interface change.  See note above.

**Review by mcfadden8 — CHANGES_REQUESTED** [review]
Overall, this looks great!

File: docs/sphinx/cookbook/dynamic_pool_heuristics.rst (hunk 12 above)
Comment (kab163) [reply docs/sphinx/cookbook/dynamic_pool_heuristics.rst:4]: I think that's a great idea because i do think this needs some updating that's probably best for a separate PR. (Especially the discussion of different heuristics.) I can make a JIRA ticket!

**Review by mcfadden8 — APPROVED** [review]

File: docs/sphinx/cookbook/dynamic_pool_heuristics.rst (hunk 12 above)
Comment (kab163) [reply docs/sphinx/cookbook/dynamic_pool_heuristics.rst:4]: See UM-905

**Review by davidbeckingsale — APPROVED** [review]
